http-proxy-configurator-operator/lib/charms/squid_forward_proxy/v0/http_proxy.py

```diff
@@ -0,0 +1,1376 @@
+
+"""Library to manage the http-proxy relation.
+
+This library contains the Requirer and Provider classes for handling the
+http-proxy interface.
+
+## Getting Started
+
+To get started using the library, you just need to fetch the library using `charmcraft`.
+
+```shell
+cd some-charm
+charmcraft fetch-lib charms.http_proxy.v0.http_proxy
+```
+
+### Using library as a requirer
+
+In the `metadata.yaml` of the charm, add the following:
+
+```yaml
+requires:
+  http-proxy:
+    interface: http-proxy
+    limit: 1
+```
+
+There are three ways to initialize the requirer class. For requirers that need to issue only
+a single request, they must either use the `HttpProxyRequirer` class or the
+`HttpProxyDynamicRequirer` class. `HttpProxyRequirer` class is used when the requirer parameters
+are known at initialization time. `HttpProxyDynamicRequirer` class is used when the requirer
+parameters are not known at initialization time and will be provided later through the
+`request_http_proxy` method. The main use case for the `HttpProxyDynamicRequirer` class
+is when the list of domains requiring proxy is dynamic. For requirers that need to issue
+multiple requests, they must use the `HttpProxyPolyRequirer` class.
+
+1. To initialize the requirer with parameters using the `HttpProxyRequirer` class:
+
+```python
+from charms.http_proxy.v0.http_proxy import {
+    HTTPProxyNotAvailableError,
+    HttpProxyRequirer
+    DEFAULT_HTTP_PROXY_INTEGRATION_NAME
+)
+
+class FooCharm(ops.CharmBase):
+    def __init__(self, *args):
+        ...
+         self.http_proxy_requirer = HttpProxyRequirer(
+            self,
+             relation_name=DEFAULT_HTTP_PROXY_INTEGRATION_NAME,
+            domains=["example.com", "example.org"],
+            auth=["userpass", "none", "srcip", "srcip+userpass"],
+            src_ips=[],
+        )
+        self.framework.observe(
+            self.on[DEFAULT_HTTP_PROXY_INTEGRATION_NAME].relation_changed, self.get_proxies
+        )
+        self.framework.observe(self.on.secret_changed, self.get_proxies)
+
+    def get_proxies(self, _: ops.EventBase):
+        try:
+            proxies = self.http_proxy_requirer.fetch_proxies()
+        except HTTPProxyUnavailableError as e:
+            logging.error(f"HTTP proxy not available. Proxy status: {e.status}")
+            return None
+        return (proxies.http_proxy, proxies.https_proxy)
+```
+
+2. To initialize the requirer with no parameters using the `HttpProxyDynamicRequirer` class:
+
+```python
+# This will simply initialize the requirer class and it won't perform any action.
+# Later provide the requirer data through the request_http_proxy method.
+
+from charms.http_proxy.v0.http_proxy import {
+    HTTPProxyNotAvailableError,
+    HttpProxyDynamicRequirer,
+    DEFAULT_HTTP_PROXY_INTEGRATION_NAME
+)
+
+class FooCharmDynamic(ops.CharmBase):
+    def __init__(self, *args):
+        ...
+        self.http_proxy_dynamic_requirer = HttpProxyDynamicRequirer(
```

Comment:
I think it'd be a better DevEx to merge the behavior of `HttpProxyRequirer` and `HttpProxyDynamicRequirer`. Is there anything preventing this? When the request is incomplete on `__init__`, just don't make the request and always allow overriding it by running `request_http_proxy`